rfc/monorepo.md

```diff
@@ -0,0 +1,557 @@
+    - A change that does not function as intended can be reverted (or fixed) quickly in a single pull request
+
+This does not come entirely for free, however:
+
+1. The build can be heavy.  There's a lot of graph for Gradle to chew through, and it can take a few minutes even with a "hot" Gradle daemon.  Cold startup is a few minutes longer, to get dependencies built once.
+    - Fortunately, Gradle caches work perfectly well with composite builds, and the cache has a *dramatic* impact.  Operating a Gradle cache is mandatory for CI (if the Gradle-GHA cache functionality is limited or insufficient), and should also be strongly considered for all development teams.
+    - It is notable that included builds can be referenced just like Gradle project syntax from the command line to drill down to specific tasks.  e.g. `./gradlew :fiat:fiat-api:test` is perfectly valid in composite-monorepo form.  This is very useful when making a smaller change that does not require waiting for a larger test suite to run locally, and Gradle is not building more than it needs to.
+1. With heavy builds comes longer CI times.  Some intelligence will need to be added to the Github Actions workflows to select Gradle tasks intelligently based on the incoming diff.
+    - We use this kind of functionality heavily in our own forks/extensions - in combination with the cache, our CI times are ~30 minutes for a full build and test of all Spinnaker code.  Code that does not touch `kork` can omit many tests, and takes between 5-15 minutes.
+1. Libraries still need to be published, even though the project itself no longer uses them - there are many external consumers.
+
+### Monorepo Github Location
+
+The monorepo will live at `github.com/spinnaker/spinnaker`.  All other repositories will be archived once the transition is completed.
+
+### Primary/Default Branch Name
+
+The primary/default branch of the monorepo will be renamed from `master` to `main`.
+
+### Yes, Deck Too
+
+This design explicitly integrates `deck` and `deck-kayenta` into the monorepo.  Cross-cutting changes do occur between the UI and the backend, and Deck is already a [Lerna](https://lerna.js.org) monorepo within itself - building all of it from its existing Gradle entry point is simple.
+
+Linking the builds of `deck`, and `deck-kayenta` is not something Gradle can do, but we have two good options here:
+
+1. Use the local `deck-kayenta` code during the build process of `deck`, using `yarn`'s workspaces feature
+1. Just integrate `deck-kayenta` as a regular `deck` module, like all the other Lerna subprojects under `packages/*`
+
+### The `spin` CLI
+
+This project is actually the most difficult to integrate into the monorepo, as golang builds tend to be really finicky about their environment.  I think Github Actions should afford us the flexibility to partition that out reasonably, but that requires some investigative work.
+
```

Comment:
I am not familiar with `goreleaser`, but definitely open to it.  GHA can use a separate container per step, so I don't actually anticipate too much trouble porting the existing process to a monorepo build, but thought it was worth a callout.

Reply:
[GoReleaser](goreleaser.com) is great. 
For the Java services we have our custom gradle publishing plugin.
For `spin` we manually wrote the artifact publishing steps in our GHA instead of taking a dependency on GoReleaser. We could change but I'm not sure what we would get.

When Gate API's are changed we need to generate swagger def and then generate `spin` code from the swagger def.
The steps are documented [here](https://github.com/spinnaker/spin/blob/master/CONTRIBUTING.md#updating-the-gate-api).
We can run a GHA to check that swagger is up to date and spin code is the same, potentially fail the build if not.
Currently we manually refresh the code from swagger when cutting a release - see [step 12 of release runbook](https://spinnaker.io/docs/releases/release-manager-runbook/#one-week-after-branches-are-cut).

Overall we can use all of our existing `spin` GHA configuration with minor changes to repo paths and similar. It will run on its own VM.